crates/red_knot_test/src/diagnostic.rs

```diff
@@ -0,0 +1,167 @@
+//! Sort and group diagnostics by line number, so they can be correlated with assertions.
+//!
+//! We don't assume that we will get the diagnostics in source order.
+
+use ruff_source_file::{LineIndex, OneIndexed};
+use ruff_text_size::Ranged;
+use std::ops::{Deref, Range};
+
+#[derive(Debug)]
+pub(crate) struct SortedDiagnostics<T> {
+    diagnostics: Vec<T>,
+    lines: Vec<LineDiagnosticRange>,
```

Comment:
maybe this would be a slightly clearer field name?

```suggestion
    line_ranges: Vec<LineDiagnosticRange>,
```